### PR TITLE
✨ Support multi-node control planes, MachineDeployments, and additional resources in the e2e framework

### PR DESCRIPTION
**Pull request by akutz**
**What this PR does / why we need it**:
This patch enhances the e2e test framework's `ControlplaneClusterInput` with support for:

* Multi-node control planes that allow the initial control plane node to come online before additional control plane nodes
* An optional `MachineDepoyment`
* Optional `AdditionalResources` (which is a slice of `runtime.Objects`)

cc @chuckha 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: test/framework/control_plane.go [hunk 1]
```diff
@@ -23,42 +23,51 @@ import (
...
-	Nodes         []Node
-	CreateTimeout time.Duration
-	DeleteTimeout time.Duration
+	Management                  ManagementCluster
+	Cluster                     *clusterv1.Cluster
+	InfraCluster                runtime.Object
+	PrimaryControlPlaneNode     Node
+	AdditionalControlPlaneNodes []Node
+	MachineDeployment           MachineDeployment
+	RelatedResources            []runtime.Object
```
Comment (chuckha) [inline test/framework/control_plane.go]: What's the use case for the related resources? Not opposed, just curious what you're using it for.

File: test/framework/control_plane.go [hunk 2]
```diff
@@ -69,81 +78,118 @@ func (input *ControlplaneClusterInput) ControlPlaneCluster() {
...
 		}
-		return err
-	}, input.CreateTimeout, 10*time.Second).Should(BeNil())
+		return nil
+	}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
+
+	By("creating related resources")
+	for _, obj := range input.RelatedResources {
+		By(fmt.Sprintf("creating a/an %s resource", obj.GetObjectKind().GroupVersionKind()))
+		Eventually(func() error {
```
Comment (chuckha) [inline test/framework/control_plane.go]: Is this to cover the case when we store related resources that are using the capi webhook servers?

File: test/framework/control_plane.go [hunk 3]
```diff
@@ -69,81 +78,118 @@ func (input *ControlplaneClusterInput) ControlPlaneCluster() {
...
-		By("creating an InfrastructureMachine resource")
-		Expect(mgmtClient.Create(ctx, node.InfraMachine)).To(Succeed())
+	// expectedNumberOfNodes is the number of nodes that should be deployed to
+	// the cluster. This is the control plane nodes plus the number of replicas
+	// defined for a possible MachineDeployment.
+	expectedNumberOfNodes := 1
 
-		By("creating a bootstrap config")
-		Expect(mgmtClient.Create(ctx, node.BootstrapConfig)).To(Succeed())
+	By("creating the primary control plane node's InfrastructureMachine resource")
```
Comment (chuckha) [inline test/framework/control_plane.go]: Cluster API should be declarative. I'd consider Cluster API to have a bug if this is not the case. I realize the test is already breaking this assumption by creating the cluster first and then the control planes etc, but in an ideal world we'd be moving toward "create everything and then wait" and not towards creating a dependency graph of objects. 

However, if this is a requirement of CAPV, I'm fine adding a new, separate test that does require an order of object applications as it is possible that not all providers have the luxury of being completely declarative.

File: test/framework/control_plane.go (hunk 1 above)
Comment (akutz) [reply test/framework/control_plane.go]: `HAProxyLoadBalancer` in my case

File: test/framework/control_plane.go (hunk 2 above)
Comment (akutz) [reply test/framework/control_plane.go]: No, this is the load balancer.

File: test/framework/control_plane.go (hunk 1 above)
Comment (chuckha) [reply test/framework/control_plane.go]: ahhh 👍 that makes sense, thanks

File: test/framework/control_plane.go (hunk 3 above)
Comment (akutz) [reply test/framework/control_plane.go]: It worked on CAPV. I think my original thinking was incorrect based on some issues I saw in-flight. I will change the PR to reflect this. Thanks!

File: test/framework/control_plane.go (hunk 2 above)
Comment (akutz) [reply test/framework/control_plane.go]: But since it’s a generic resource, I figure an eventually makes sense.

**Comment by akutz** [comment]
Okay @chuckha, I should have addressed your concern around ordering.

File: test/framework/control_plane.go (hunk 3 above)
Comment (akutz) [reply test/framework/control_plane.go]: This should be resolved now, thanks!

File: test/framework/control_plane.go [hunk 4]
```diff
@@ -69,81 +77,109 @@ func (input *ControlplaneClusterInput) ControlPlaneCluster() {
...
-				Namespace: node.Machine.GetNamespace(),
-				Name:      node.Machine.GetName(),
-			}
-			if err := mgmtClient.Get(ctx, key, machine); err != nil {
-				return err.Error()
-			}
-			return machine.Status.Phase
-		}, input.CreateTimeout, 10*time.Second).Should(Equal(string(clusterv1.MachinePhaseRunning)))
+	// Create the machine deployment if the replica count >0.
+	if machineDeployment := input.MachineDeployment.MachineDeployment; machineDeployment != nil {
```
Comment (chuckha) [inline test/framework/control_plane.go]: I'm ok with this for now, but I want us to watch out for making this test do too much. This if statement has the ability to quickly grow into other Ifs and muddy this test. If there's a set up step we can establish and reuse that between different tests (control plane, machine deployment, etc), that would be great, but that's longterm planning and I'm totally cool with this going through but acknowledge we should watch out for adding too many ifs in this test.

Comment (akutz) [reply test/framework/control_plane.go]: Hmmm, I kind of see your point? But I think what you're saying makes more sense before when the inputs were separate from the function that used them. In that design the inputs could be viewed as immutable, or at least immutable once copied by value into the functions that used them. 

In the refactor from https://github.com/kubernetes-sigs/cluster-api/commit/3efd98f2be0d39036971d602b2be2ebe7753aa4c I think it makes sense to support optional resources, of which a `MachineDeployment` is one such thing. Because now the input is kind of the progenitor of the resource that performs the actual test.

In other words, I liked that the functions for `ControlPlane` or `MultiNodeControlPlane`, etc. were functional in design. You could then design them to accept multiple input types that could be various CAPI deployment strategies. However, having one input that can do *everything*, even if optionally, seems to be in line with the current model.
